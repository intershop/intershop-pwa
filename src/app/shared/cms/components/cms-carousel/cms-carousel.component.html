<div *ngIf="pagelet" [ngClass]="pagelet.stringParam('CSSClass')">
  <ish-content-slot [wrapper]="true" [slot]="'app_sf_base_cm:slot.carousel.items.pagelet2-Slot'" [pagelet]="pagelet">
    <div [ngClass]="pagelet.stringParam('CarouselCSSClass')">
      <ngb-carousel
        #ngbCarousel
        [interval]="0"
        [showNavigationArrows]="pageletSlides.length > 1"
        [showNavigationIndicators]="pagelet.booleanParam('ShowIndicators')"
      >
        <ng-template ngbSlide *ngFor="let slidePagelets of pageletSlides">
          <div class="row">
            <ng-container *ngFor="let slidePagelet of slidePagelets">
              <div [ngClass]="'col-sm-' + itemGridSize">
                <ish-content-pagelet [pageletId]="slidePagelet"></ish-content-pagelet>
              </div>
            </ng-container>
          </div>
        </ng-template>
      </ngb-carousel>
    </div>
  </ish-content-slot>
</div>
