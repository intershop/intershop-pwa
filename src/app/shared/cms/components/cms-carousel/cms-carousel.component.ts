import {
  ApplicationRef,
  ChangeDetectionStrategy,
  Component,
  Input,
  OnChanges,
  OnDestroy,
  ViewChild,
} from '@angular/core';
import { NgbCarousel } from '@ng-bootstrap/ng-bootstrap';
import { Subject } from 'rxjs';
import { mapTo, take, takeUntil } from 'rxjs/operators';

import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { arraySlices } from 'ish-core/utils/functions';
import { whenTruthy } from 'ish-core/utils/operators';
import { CMSComponent } from 'ish-shared/cms/models/cms-component/cms-component.model';

@Component({
  selector: 'ish-cms-carousel',
  templateUrl: './cms-carousel.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CMSCarouselComponent implements CMSComponent, OnChanges, OnDestroy {
  @Input() pagelet: ContentPageletView;

  @ViewChild('ngbCarousel') carousel: NgbCarousel;

  slideItems = 6;
  itemGridSize = 12;
  pageletSlides: string[][] = [];

  constructor(private appRef: ApplicationRef) {}

  private destroy$ = new Subject();

  ngOnChanges() {
    if (this.pagelet.hasParam('SlideItems')) {
      this.slideItems = this.pagelet.numberParam('SlideItems');
    }
    this.itemGridSize = (12 - (12 % this.slideItems)) / this.slideItems;

    const slotPagelets = this.pagelet.slot('app_sf_base_cm:slot.carousel.items.pagelet2-Slot').pageletIDs;
    this.pageletSlides = arraySlices(slotPagelets, this.slideItems);

    this.appRef.isStable
      .pipe(
        whenTruthy(),
        mapTo(this.pagelet.booleanParam('StartCycling') ? this.pagelet.numberParam('SlideInterval', 5000) : 0),
        take(1),
        takeUntil(this.destroy$)
      )
      .subscribe(val => {
        if (val) {
          this.carousel.interval = val;
          this.carousel.cycle();
        }
      });
  }

  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }
}
