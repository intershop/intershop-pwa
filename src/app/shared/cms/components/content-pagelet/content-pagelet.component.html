<ish-content-design-view-wrapper [pageletId]="pageletId">
  <ng-container #cmsOutlet />
</ish-content-design-view-wrapper>
