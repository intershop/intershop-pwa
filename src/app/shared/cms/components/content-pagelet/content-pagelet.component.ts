import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ComponentFactoryResolver,
  Injector,
  Input,
  OnChanges,
  OnDestroy,
  OnInit,
  SimpleChange,
  ViewChild,
  ViewContainerRef,
} from '@angular/core';
import { ReplaySubject, Subject } from 'rxjs';
import { switchMap, takeUntil } from 'rxjs/operators';

import { CMSFacade } from 'ish-core/facades/cms.facade';
import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { whenTruthy } from 'ish-core/utils/operators';
import { CMSComponentProvider, CMS_COMPONENT } from 'ish-shared/cms/configurations/injection-keys';
import { CMSComponent } from 'ish-shared/cms/models/cms-component/cms-component.model';
import { SfeAdapterService } from 'ish-shared/cms/sfe-adapter/sfe-adapter.service';
import { SfeMetadataWrapper } from 'ish-shared/cms/sfe-adapter/sfe-metadata-wrapper';
import { SfeMapper } from 'ish-shared/cms/sfe-adapter/sfe.mapper';

/**
 * The Content Pagelet Container Component renders the pagelet for the given 'pageletId'.
 * For the rendering an Angular component is used that has to be provided
 * for the DefinitionQualifiedName of the pagelet.
 *
 * @example
 * <ish-content-pagelet [pageletId]="pagelet"></ish-content-pagelet>
 */
@Component({
  selector: 'ish-content-pagelet',
  templateUrl: './content-pagelet.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ContentPageletComponent extends SfeMetadataWrapper implements OnChanges, OnInit, OnDestroy {
  /**
   * The Id of the Pagelet that is to be rendered.
   */
  @Input() pageletId: string;

  @ViewChild('cmsOutlet', { read: ViewContainerRef, static: true }) cmsOutlet: ViewContainerRef;

  private pageletId$ = new ReplaySubject<string>(1);
  private destroy$ = new Subject();

  constructor(
    private injector: Injector,
    private componentFactoryResolver: ComponentFactoryResolver,
    private sfeAdapter: SfeAdapterService,
    private cmsFacade: CMSFacade,
    private cdRef: ChangeDetectorRef
  ) {
    super();
  }

  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }

  ngOnInit() {
    this.pageletId$
      .pipe(
        switchMap(pageletId => this.cmsFacade.pagelet$(pageletId)),
        whenTruthy(),
        takeUntil(this.destroy$)
      )
      .subscribe(pagelet => {
        this.mapComponent(pagelet);
        if (this.sfeAdapter.isInitialized()) {
          this.setSfeMetadata(SfeMapper.mapPageletViewToSfeMetadata(pagelet));
        }
        this.cdRef.markForCheck();
      });
  }

  ngOnChanges() {
    this.pageletId$.next(this.pageletId);
  }

  private mapComponent(pagelet: ContentPageletView) {
    const components = this.injector.get<CMSComponentProvider[]>(CMS_COMPONENT, []);
    const mappedComponent = components.find(c => c.definitionQualifiedName === pagelet.definitionQualifiedName);

    if (mappedComponent) {
      const componentRef = this.createComponent(mappedComponent);
      this.initializeComponent(componentRef.instance, pagelet);
    } else {
      console.warn(`did not find mapping for ${pagelet.id} (${pagelet.definitionQualifiedName})`);
    }
  }

  private createComponent(mappedComponent: CMSComponentProvider) {
    const factory = this.componentFactoryResolver.resolveComponentFactory(mappedComponent.class);
    this.cmsOutlet.clear();
    return this.cmsOutlet.createComponent(factory);
  }

  private initializeComponent(instance: CMSComponent, pagelet: ContentPageletView) {
    instance.pagelet = pagelet;

    // OnChanges has to be manually invoked on dynamically created components
    if (instance.ngOnChanges) {
      instance.ngOnChanges({ pagelet: new SimpleChange(undefined, pagelet, true) });
    }
  }
}
