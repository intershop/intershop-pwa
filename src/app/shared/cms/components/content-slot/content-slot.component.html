<ng-content *ngIf="wrapper; else renderPagelets"></ng-content>

<ng-template #renderPagelets>
  <ish-content-pagelet
    *ngFor="let pagelet of pagelet.slot(slot).pageletIDs"
    [pageletId]="pagelet"
    [callParameters]="callParameters"
  ></ish-content-pagelet>
</ng-template>
