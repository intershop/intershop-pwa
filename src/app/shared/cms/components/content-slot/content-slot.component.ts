import { ChangeDetectionStrategy, Component, Input, OnChanges } from '@angular/core';

import { ContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { SfeAdapterService } from 'ish-shared/cms/sfe-adapter/sfe-adapter.service';
import { SfeMetadataWrapper } from 'ish-shared/cms/sfe-adapter/sfe-metadata-wrapper';
import { SfeMapper } from 'ish-shared/cms/sfe-adapter/sfe.mapper';

/**
 * The Content Slot Container Component renders the assigned sub pagelets
 * of the identified 'slot' of the given 'pagelet'.
 * By default it is just using the {@link ContentPageletContainerComponent}
 * to render each sub pagelet.
 * If more specific HTML or functionality is needed for the rendering
 * the 'wrapper' flag needs to be used to use the provided HTML for rendering.
 *
 * @example
 * <ish-content-slot [slot]="'app_pwa:slot.pagelet2-Slot'" [pagelet]="pagelet"></ish-content-slot>
 *
 * <ish-content-slot [wrapper]="true" [slot]="'app_pwa:slot.pagelet2-Slot'" [pagelet]="pagelet">
 *               <div *ngFor="let slotPagelet of slotPagelets">
 *                 <ish-content-pagelet [pageletId]="slotPagelet"></ish-content-pagelet>
 *               </div>
 * </ish-content-slot>
 */
@Component({
  selector: 'ish-content-slot',
  templateUrl: './content-slot.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ContentSlotComponent extends SfeMetadataWrapper implements OnChanges {
  /**
   * The DefinitionQualifiedName of the slot that should be rendered.
   */
  @Input() slot: string;
  /**
   * The parent Pagelet that contains the slot.
   */
  @Input() pagelet: ContentPageletView;
  /**
   * An optional flag that controls the rendering of the pagelets with the wrapped HTML content.
   */
  @Input() wrapper?: boolean;

  constructor(private sfeAdapter: SfeAdapterService) {
    super();
  }

  ngOnChanges() {
    if (this.pagelet && this.sfeAdapter.isInitialized()) {
      this.setSfeMetadata(SfeMapper.mapSlotViewToSfeMetadata(this.pagelet, this.slot));
    }
  }
}
