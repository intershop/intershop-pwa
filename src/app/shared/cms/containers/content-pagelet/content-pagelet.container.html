<ng-container #cmsOutlet></ng-container>
