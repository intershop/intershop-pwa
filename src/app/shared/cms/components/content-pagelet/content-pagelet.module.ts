import { Injector, NgModule } from '@angular/core';

import { ModuleLoaderService } from 'ish-core/utils/module-loader/module-loader.service';
import { ContentPageletComponent } from 'ish-shared/cms/components/content-pagelet/content-pagelet.component';

const exportedComponents = [ContentPageletComponent];

@NgModule({
  declarations: [...exportedComponents],
  exports: [...exportedComponents],
})
export class ContentPageletModule {
  constructor(moduleLoader: ModuleLoaderService, injector: Injector) {
    moduleLoader.init(injector);
  }
}
