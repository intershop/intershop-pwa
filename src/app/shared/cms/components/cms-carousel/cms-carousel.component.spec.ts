import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { NgbCarouselModule } from '@ng-bootstrap/ng-bootstrap';
import { MockComponent } from 'ng-mocks';

import { createContentPageletView } from 'ish-core/models/content-view/content-view.model';
import { ContentPageletComponent } from 'ish-shared/cms/components/content-pagelet/content-pagelet.component';
import { ContentSlotComponent } from 'ish-shared/cms/components/content-slot/content-slot.component';

import { CMSCarouselComponent } from './cms-carousel.component';

describe('Cms Carousel Component', () => {
  let component: CMSCarouselComponent;
  let fixture: ComponentFixture<CMSCarouselComponent>;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [CMSCarouselComponent, MockComponent(ContentPageletComponent), MockComponent(ContentSlotComponent)],
      imports: [NgbCarouselModule],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CMSCarouselComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;
    const pagelet = {
      id: 'id',
      definitionQualifiedName: 'fq',
      domain: 'domain',
      displayName: 'pagelet',
      configurationParameters: {
        CSSClass: 'foo-class',
        SlideItems: 2,
      },
      slots: [
        {
          displayName: 'test',
          definitionQualifiedName: 'app_sf_base_cm:slot.carousel.items.pagelet2-Slot',
          pageletIDs: ['slide1', 'slide2'],
        },
      ],
    };
    component.pagelet = createContentPageletView(pagelet);
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => component.ngOnChanges()).not.toThrow();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });
});
