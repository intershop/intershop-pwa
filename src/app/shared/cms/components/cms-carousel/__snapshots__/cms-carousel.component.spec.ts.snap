// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cms Carousel Component should be created 1`] = `
<div class="foo-class" ng-reflect-ng-class="foo-class">
  <ish-content-slot ng-reflect-slot="app_sf_base_cm:slot.carousel.i" ng-reflect-wrapper="true"
    ><div>
      <ngb-carousel
        class="carousel slide"
        tabindex="0"
        ng-reflect-interval="0"
        ng-reflect-show-navigation-arrows="false"
        style="display: block;"
        ><div class="carousel-inner">
          <div class="carousel-item active">
            <div class="row">
              <div class="col-sm-6" ng-reflect-ng-class="col-sm-6">
                <ish-content-pagelet ng-reflect-pagelet-id="slide1"></ish-content-pagelet>
              </div>
              <div class="col-sm-6" ng-reflect-ng-class="col-sm-6">
                <ish-content-pagelet ng-reflect-pagelet-id="slide2"></ish-content-pagelet>
              </div>
            </div>
          </div></div
      ></ngb-carousel></div
  ></ish-content-slot>
</div>
`;
